README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# BayesianLaterality

<!-- badges: start -->

[![CRAN_Status_Badge](https://www.r-pkg.org/badges/version/BayesianLaterality)](https://cran.r-project.org/package=BayesianLaterality)
[![R-CMD-check](https://github.com/LCBC-UiO/BayesianLaterality/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/LCBC-UiO/BayesianLaterality/actions/workflows/R-CMD-check.yaml)
<!-- badges: end -->

BayesianLaterality contains a function for predicting latent hemispheric
dominance based on observed laterality using a Bayes classifier
developed by [Sørensen and Westerhausen
(2020)](https://doi.org/10.1080/1357650X.2020.1769124). See also the
[accompanying Shiny
app](https://osorensen.shinyapps.io/BayesianLateralityApp/).

## Installation

You can install BayesianLaterality from
[CRAN](https://cran.r-project.org/web/packages/BayesianLaterality/)
with:

``` r
install.packages("BayesianLaterality")
```

Install the latest development version from
[GitHub](https://github.com/LCBC-UiO/BayesianLaterality) with:

``` r
# install.packages("remotes")
remotes::install_github("LCBC-UiO/BayesianLaterality")
```

## Application Example

``` r
library(BayesianLaterality)
```

The main (and only) function of the package is `predict_dominance()`. To
see the arguments that can be set by the user and a more extended
example, type `?predict_dominance` in the R terminal. Here is a simple
example. The dataset `example_data1` contains three laterality
measurements on three right-handed individuals.

``` r
example_data1
#>   listening handedness
#> 1        20      right
#> 2        23      right
#> 3        14      right
```

We then obtain predicted hemispheric dominance as follows. The ID column
reflects the row in the original dataset.

``` r
predict_dominance(example_data1)
#> No ID column in data, assuming one subject per row.
#> # A tibble: 9 × 4
#>   ID    handedness dominance probability
#>   <chr> <chr>      <chr>           <dbl>
#> 1 1     right      left          0.994  
#> 2 1     right      right         0.00583
#> 3 1     right      none          0      
#> 4 2     right      left          0.996  
#> 5 2     right      right         0.00402
#> 6 2     right      none          0      
#> 7 3     right      left          0.988  
#> 8 3     right      right         0.0122 
#> 9 3     right      none          0
```


news.md
# BayesianLaterality version 0.1.2

Update package documentation to use the new special sentinel "_PACKAGE".

# BayesianLaterality version 0.1.1

Updated DESCRIPTION so Authors(at)R syntax is being used, and added URL.

# BayesianLaterality version 0.1.0

This is the first release of the package.


test_data.R

data <- data.frame(
  listening = c(-20, -23, -14),
  handedness = "left",
  stringsAsFactors = T
)
predict_asymmetry(data)

parameters = dplyr::tibble(
  dominance = rep(c("left", "right", "none"), each = 2),
  handedness = rep(c("left", "right"), 3),
  mean_li = c(10, 12, -24, -24, 0, 0),
  sd_li = rep(22, 6),
  prob_dominance = c(.65, .87, .20, .04, .15, .09)
)
truncation = c(-100, 100)
icc = NULL


diag(3) *(1-.22) + .22


library(tidyverse)
n <- 100
reps <- 3

data <- tibble(
  ID = 1:n,
  ID_mean = truncnorm::rtruncnorm(n, a = 0, b = 100, mean = 10, sd = 10),
  handedness = "left"
) %>%
  mutate(
    listening = map(ID_mean, ~ rnorm(reps, mean = .x, sd = 10))
  ) %>%
  unnest(listening)

predict_asymmetry(data)


R/predict_asymmetry.R
#' Predict hemispheric dominance
#'
#' @param data Data frame with the following columns:
#' \itemize{
#' \item \code{listening}: Score between -100 and 100.
#' \item \code{handedness}: \code{"left"} for adextral (non-right-handed) and
#'       \code{"right"} for dextral (right-handed)
#'       }
#'  In addition, an optional column name \code{ID}
#'   can be provided, giving the subject ID. If a subject has multiple
#'   measurements, the posterior based on all measurements is provided. If the
#'   \code{ID} column is missing, each row is assumed to be measured on a
#'   separate subject.
#'
#' @param parameters Data frame in which the first two columns specify combinations
#' of hemispheric dominance and handedness and the last three columns specify
#' the corresponding parameter values. In particular, the columns are defined as follows:
#' \itemize{
#' \item \code{dominance}: character specifying hemispheric dominance.
#' \item \code{handedness}: character specifying handedness.
#' \item \code{mean_li}: mean dichotic listening score.
#' \item \code{sd_li}: standard deviation of dichotic listening score.
#' \item \code{prob_dominance}: probability of hemispheric dominance given handedness.
#' }
#' @param truncation Numeric vector with two elements specifying the lower and upper
#' bounds for truncation of the normal distribution for dichotic listening scores.
#' @param icc Intraclass correlation for repeated measurements on the same individual.
#' Defaults to 0.
#'
#' @return The probability of left or right hemispheric dominance in additional
#'   columns of \code{data}.
#' @export
#' @examples
#' ## Simple test dataset
#' data <- data.frame(
#'           listening = c(-20, -23, -14),
#'           handedness = "left",
#'           stringsAsFactors = FALSE
#'          )
#' ## Compute predictions
#' predict_dominance(data)
#'
#' ## More interesting example, with multiple measurements per individual.
#' library(dplyr); library(purrr); library(tidyr); library(truncnorm)
#' ## First we sample test data
#' n <- 100 # number of individuals
#' reps <- 3 # number of measurements per individual
#' ## The distribution of subject means has standard deviation 10, and the
#' ## actual measurements for each subject are distributed with a standard
#' ## deviation of 10 around this mean.
#' set.seed(234)
#' data <- tibble(
#'                ID = factor(1:n),
#'                subject_mean = rtruncnorm(n, a = 0, b = 100, mean = 10, sd = 10),
#'                handedness = "left") %>%
#'   mutate(
#'     listening = map(subject_mean, ~ rtruncnorm(reps, a = -100, b = 100,
#'                     mean = .x, sd = 10))
#'   ) %>%
#'   unnest(listening)
#'
#' predict_dominance(data)
#'
#'
predict_dominance <- function(data,
                              parameters = dplyr::tibble(
                                dominance = rep(c("left", "right", "none"), each = 2),
                                handedness = rep(c("left", "right"), 3),
                                mean_li = c(10, 12, -24, -24, 0, 0),
                                sd_li = c(24.9, 17.0, 24.9, 17.0, 22, 22),
                                prob_dominance = c(.65, .87, .35, .13, 0, 0)
                              ),
                              truncation = c(-100, 100),
                              icc = 0
                              ){

  stopifnot(icc >= -1 && icc <= 1)

  # Check if data contains an ID column
  if(!"ID" %in% colnames(data)) {
    message("No ID column in data, assuming one subject per row.")
    data$ID = as.character(seq(1, nrow(data), by = 1))
  }

  dat1 <- dplyr::select(data, .data$ID, .data$listening, .data$handedness)
  dat1 <- dplyr::inner_join(dat1, parameters, by = "handedness")
  dat1 <- dplyr::select(dat1, .data$ID, .data$handedness, .data$dominance,
                        .data$prob_dominance, .data$mean_li, .data$sd_li, .data$listening)
  dat2 <- tidyr::nest(dat1, df = c(.data$listening, .data$mean_li, .data$sd_li))
  dat3 <- dplyr::mutate(dat2,
                log_prob_listening = purrr::map_dbl(.data$df, function(x) {
                  tmvtnorm::dtmvnorm(x$listening,
                                     mean = x$mean_li,
                                     sigma = x$sd_li^2 * (diag(nrow(x)) * (1 - icc) + icc),
                                     lower = rep(truncation[[1]], nrow(x)),
                                     upper = rep(truncation[[2]], nrow(x)),
                                     log = TRUE
                                     )
                }),
                log_posterior = log(.data$prob_dominance) + .data$log_prob_listening,
                probability = exp(.data$log_posterior)
                )
  dat4 <- dplyr::group_by(dat3, .data$ID)
  dat5 <- dplyr::mutate(dat4, probability = .data$probability / sum(.data$probability))

  dplyr::select(dplyr::ungroup(dat5), .data$ID, .data$handedness,
                .data$dominance, .data$probability)
}



R/predict_dominance.R
#' Predict hemispheric dominance based on observed laterality measures, using the methods
#' described in \insertCite{Sorensen2020;textual}{BayesianLaterality}.
#'
#' @param data Data frame with the following columns:
#' \itemize{
#' \item \code{listening}: Score between -100 and 100.
#' \item \code{handedness}: \code{"left"} for adextral (non-right-handed) and
#'       \code{"right"} for dextral (right-handed)
#'       }
#'  In addition, an optional column name \code{ID}
#'   can be provided, giving the subject ID. If a subject has multiple
#'   measurements, the posterior based on all measurements is provided. If the
#'   \code{ID} column is missing, each row is assumed to be measured on a
#'   separate subject.
#'
#' @param parameters Data frame in which the first two columns specify combinations
#' of hemispheric dominance and handedness and the last three columns specify
#' the corresponding parameter values. In particular, the columns are defined as follows:
#' \itemize{
#' \item \code{dominance}: character specifying hemispheric dominance.
#' \item \code{handedness}: character specifying handedness.
#' \item \code{mean_li}: mean dichotic listening score.
#' \item \code{sd_li}: standard deviation of dichotic listening score.
#' \item \code{prob_dominance}: probability of hemispheric dominance given handedness.
#' }
#' @param truncation Numeric vector with two elements specifying the lower and upper
#' bounds for truncation of the normal distribution for dichotic listening scores.
#' @param icc Intraclass correlation for repeated measurements on the same individual.
#' Defaults to 0.
#'
#' @return The probability of left or right hemispheric dominance in additional
#'   columns of \code{data}.
#' @export
#' @examples
#' # The package comes with two example datasets.
#' # The first contains single measurements on three subjects.
#' # We can first take a look at the data
#' example_data1
#' # Next, compute predictions.
#' # Since there is no ID column, predict_dominance() will print a message telling
#' # the user that the rows are assumed to contain observations from different subjects.
#' predict_dominance(example_data1)
#'
#' # The next example dataset contains repeated measurements
#' example_data2
#'
#' # We compute the predictions as before:
#' predict_dominance(example_data2)
#'
#' @references
#' \insertAllCited{}
#'
#' @importFrom rlang .data
#' @importFrom Rdpack reprompt
predict_dominance <- function(data,
                              parameters = dplyr::tibble(
                                dominance = rep(c("left", "right", "none"), each = 2),
                                handedness = rep(c("left", "right"), 3),
                                mean_li = c(10, 12, -24, -24, 0, 0),
                                sd_li = c(24.9, 17.0, 24.9, 17.0, 22, 22),
                                prob_dominance = c(.65, .87, .35, .13, 0, 0)
                              ),
                              truncation = c(-100, 100),
                              icc = 0
                              ){

  stopifnot(icc >= -1 && icc <= 1)

  # Check if data contains an ID column
  if(!"ID" %in% colnames(data)) {
    message("No ID column in data, assuming one subject per row.")
    data$ID = as.character(seq(1, nrow(data), by = 1))
  }

  dat1 <- dplyr::select_at(data, dplyr::vars("ID", "listening", "handedness"))
  dat1 <- dplyr::inner_join(dat1, parameters, by = "handedness")
  dat1 <- dplyr::select_at(dat1, dplyr::vars("ID", "handedness", "dominance",
                                             "prob_dominance", "mean_li", "sd_li", "listening"))
  dat2 <- tidyr::nest(dat1, df = c("listening", "mean_li", "sd_li"))
  dat3 <- dplyr::mutate(dat2,
                log_prob_listening = purrr::map_dbl(.data$df, function(x) {
                  tmvtnorm::dtmvnorm(x$listening,
                                     mean = x$mean_li,
                                     sigma = x$sd_li^2 * (diag(nrow(x)) * (1 - icc) + icc),
                                     lower = rep(truncation[[1]], nrow(x)),
                                     upper = rep(truncation[[2]], nrow(x)),
                                     log = TRUE
                                     )
                }),
                log_posterior = log(.data$prob_dominance) + .data$log_prob_listening,
                probability = exp(.data$log_posterior)
                )
  dat4 <- dplyr::group_by_at(dat3, dplyr::vars("ID"))
  dat5 <- dplyr::mutate_at(dat4, dplyr::vars("probability"), ~ . / sum(.))

  dplyr::select_at(dplyr::ungroup(dat5), dplyr::vars("ID", "handedness",
                                                     "dominance", "probability"))
}

